File: top-interview-150/python/003_longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        str_len = len(s)
        if str_len == len(set(s)):
            return str_len

        max_window, t_set_len = 1, 1
        for i in range(0, str_len):
            for w in range(max_window, str_len):
                if (i+w) <= str_len:
                    t = s[i:i+w]
                else:
                    break
                t_set_len = len(set(t))
                if (len(t) != t_set_len):
                    break
            if t_set_len > max_window:
                max_window = t_set_len
        return max_window

File: top-interview-150/c++/003_longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len_s = s.size();
        unordered_set<char> s_set(s.begin(), s.end());
        int len_s_set = s_set.size();
        if (len_s == len_s_set) return len_s;

        int max_window = 1, len_t_set = 1;
        //cout << "max_window: " << max_window << " len_t_set: " << len_t_set << endl;
        for (int i=0; i<len_s; i++) {
            if (len_t_set > max_window) max_window = len_t_set;
            if (i+max_window > len_s) break;
            for (int w=max_window; w<len_s; w++) {
                //cout << "i: " << i << " w: " << w << endl;
                if ((i+w) > len_s) break;
                else {
                    set<char> t(s.begin()+i, s.begin()+i+w);
                    //cout << string(t.begin(), t.end()) << endl;
                    len_t_set = t.size();
                    if (w != len_t_set) break;
                }
            }
        }
        return max_window;
    }
};
